docs/MUSIC_COMPOSITION_ENVIRONMENT.md

```diff
@@ -0,0 +1,661 @@
+# Music Composition Environment
+
+## Vision
+
+Enable Cadenza to power an algorithmic music composition system that combines the best of live coding, offline rendering, and precise musical control. The system integrates with the [bach](https://github.com/camshaft/bach) discrete event simulator and [euphony](https://github.com/camshaft/euphony-rs) project to create a comprehensive music programming environment.
+
+Key capabilities:
+1. **Live Performance Mode**: Real-time execution responding to MIDI/UI events
+2. **Offline Rendering**: Non-realtime parallel rendering as fast as possible
+3. **DSP Precompilation**: Performance-critical audio processing compiled to native code
+4. **Interpreted Control**: High-level composition logic interpreted for flexibility
+5. **Autonomous Tasks**: Multiple concurrent tasks coordinating via discrete event simulation
+6. **Interactive REPL**: Live experimentation and parameter adjustment
+
+## Goals
+
+### Primary Goals
+
+1. **Dual Execution Modes**
+   - **Online (Live)**: Real-time event handling with predictable latency
+   - **Offline (Render)**: Fast parallel rendering without real-time constraints
+   - Switch between modes seamlessly
+   - Share composition code between both modes
+
+2. **Hybrid Compilation Strategy**
+   - Precompile DSP algorithms to native code for performance
+   - Interpret control logic for flexibility and live updates
+   - JIT compilation for hot paths when beneficial
+   - Zero-cost abstractions where possible
+
+3. **Discrete Event Simulation**
+   - Integration with bach library for event scheduling
+   - Autonomous tasks emitting commands and MIDI events
+   - Precise timing control with musical time units
+   - Task coordination and synchronization
+
+4. **Interactive Development**
+   - REPL for live coding and parameter tweaking
+   - Hot-reload of composition changes
+   - Visual feedback (waveforms, spectrograms, parameter graphs)
```

Comment:
This is huge. Ideally we'd be able to inspect anything and have insight into what is happening in the environment so it's easier to iterate and understand how things are being composed.